### ISSUE TITLE
Wait for resources to be created

### ISSUE THREAD

**Issue by ale8k** [juju/terraform-provider-juju#116]
Currently the provider provisions resources and considers that job done. It would be nice if the terraform provider waited until a desired state (ideally running and idle/active). I managed to simulate this with local-exec [via juju wait-for] provisioners and a create_before_destroy lifecycle hook. 

It can be found here: https://github.com/canonical/livepatch-server/pull/900/files

The problem is the destroy provisioners don't run when using create_before_destroy... 

So it would be a great addition that:
1. I apply my plan
2. The provider expects resources to be created before destroy
3. Resource X fails
4. Terraform removes X and updates in place all resources prior OR terraform removes X and all other created before destruction resources, such that it is just considered a "failed deploy" from the users perspective, but the existing resources and state are safe.

Is this a feature that could be added / makes sense?

Thanks

**Comment by juanmanuel-tirado** [comment]
You're right. The current implementation is particularly naive and it only uses the Juju API to materialize a given plan. Additional control support sounds reasonable and it would probably be very useful in many scenarios. Thoughts about this @jnsgruk ? I will add it to the wishlist for further consideration.

**Comment by jnsgruk** [comment]
This will likely need some quite careful design, and defining what is a success for a deploy will need some care, I think

**Comment by mthaddon** [comment]
I'd agree we should be careful in defining what a successful deployment is. Many charms don't go to active/idle state as they may block waiting on relations or particular configuration, for instance.

Also, just to mention that I've just tried deploying a primary charm (vault) and three subordinates (landscape-client, telegraf, ubuntu-advantage) and two of the subordinates failed to deploy as far as I can tell because the primary application didn't exist yet: https://paste.ubuntu.com/p/w3SfCKNMr7/. Rerunning the same `apply` worked fine, since the vault application was found at that point.

**Comment by mthaddon** [comment]
Can this not be implemented by using the new `juju wait-for` functionality (would be Juju 3 only, but still worthwhile)? https://juju.is/docs/olm/juju-wait-for

**Comment by hmlanigan** [comment]
Another consideration are subordinate charms, which are not fully deployed until a relation is added and the principal charm has running units. Thus depending on other terraform resources to be complete and done also.

Are terraform provider expectations documented somewhere? Is the expectation that the apply command completes that the changes are fully made and finished? Or that the changes are started?

We do have some timing windows to solve such as #272

**Comment by SimoneDutto** [comment]
https://github.com/juju/terraform-provider-juju/pull/742

In this PR we've fixed this issue of waiting for application's destroy.
